Feature/162396232 redesign ea homepage

lingyun1010 opened the pull request:
Define a properties file that can be read/written by data production (`species/atlas-information.json`).
Create an endpoint and a new card type/factory for gxa popular species.
Use [homepage cards container](https://github.com/ebi-gene-expression-group/atlas-homepage-cards/pull/9) and [homapge panel contrainer](https://github.com/ebi-gene-expression-group/scxa-home-landing-page/pull/1).
Make featured experiments a single row with 6 cards.
Explicit baseline/differential experiments, don’t use B/D circles.

lingyun1010 committed:
Add atlas-homepage-card into gxa bundles

lingyun1010 committed:
Update css style for homepage and b/d full texts

lingyun1010 committed:
Change css style and remove view-all button

lingyun1010 committed:
Add home page container bundle in gxa

lingyun1010 committed:
Update props for homepage panel container in SC

lingyun1010 committed:
Add a json controller for popular species in GXA

lingyun1010 committed:
Update bundles

lingyun1010 committed:
Use homepage cards for popular  species in gxa

lingyun1010 committed:
Rename the homepage panel bundles

lingyun1010 committed:
Build bundles

lingyun1010 committed:
Remove useless props

lingyun1010 committed:
Add a cardmodel factory for atlas homepage species

cards

lingyun1010 committed:
Add a url helper for differencial and baseline url

lingyun1010 committed:
Use the atlas homepage card factory

lingyun1010 committed:
Add a new card type

lingyun1010 committed:
Read atlas information json file

lingyun1010 committed:
Add atlas info attribute to homepage jsp

lingyun1010 committed:
Update homepage hero with info from json file

lingyun1010 committed:
Update homepage card bundle with new export

lingyun1010 committed:
Change featured experiment name and css style

lingyun1010 committed:
Add a test for jsonPopularSpecies Controller

lingyun1010 committed:
Clean up code

lingyun1010 committed:
Add a test for new popular species cardtype factory

monicajianu commented:
@lingyun1010 Where can I find a sample `species/atlas-information.json` file? I can't run the application without it.

monicajianu commented on base/src/main/java/uk/ac/ebi/atlas/model/card/CardIconType.java:
```diff
@@ -1,5 +1,5 @@
 
 public enum CardIconType {
-    SPECIES, IMAGE
+    SPECIES, IMAGE, ATLASSPECIES
```
How is `ATLASSPECIES` different to the `SPECIES` card type?

lingyun1010 replied:
atlas species card has differential and baseline button

monicajianu replied:
This enum does not denote the card type, but the type of icon used in the card. I.e. a `SPECIES` icon type is represented by the EBI species icons, and a `IMAGE` icon type will make use of static image assets such as `.png` files. How is the `ATLASSPECIES` icon type different?

lingyun1010 replied:
Because we need to render different type of cards when we received the card type from backend. `ATLASSPECIES` specified that it needs a species card with B and D links, which is different from the rendering of `SPECIES` cards.

lingyun1010 replied:
I have made corresponding changes to use the `SPECIES` card type only in gxa and sc homepages.
![Atlas homepage screenshot](https://user-images.githubusercontent.com/33519183/51549707-ceffa700-1e62-11e9-9082-b0fcc334432a.png)

monicajianu commented on base/src/main/java/uk/ac/ebi/atlas/model/card/CardModelFactory.java:
```diff
@@ -32,6 +33,24 @@ public CardModel create(PopularSpeciesInfo popularSpeciesInfo) {
+        List content = new ArrayList();
+        content.add(Pair.of(popularSpeciesInfo.totalExperiments() + " experiments",
+                Optional.empty()));
+        content.add(Pair.of(popularSpeciesInfo.baselineExperiments() + " ",
```
You should add the words "baseline" and "differential" after the number of experiments, rather than just an empty space. Seeing as the B and D icons have been removed, the numbers on their own don't make much sense!

lingyun1010 replied:
I think you have misunderstood this branch. The design of `ATLASSPECIES` card is [homepage cards container](https://github.com/ebi-gene-expression-group/atlas-homepage-cards/pull/9), where you can see the number is behind `Baseline` and `Differential` icons. This atlas branch is just used to pass the json file for cards container for now, but have not bundled the unreviewed [homepage cards container](https://github.com/ebi-gene-expression-group/atlas-homepage-cards/pull/9).

monicajianu commented on base/src/main/java/uk/ac/ebi/atlas/model/card/CardModelFactory.java:
```diff
@@ -32,6 +33,24 @@ public CardModel create(PopularSpeciesInfo popularSpeciesInfo) {
 
+
+    public CardModel createAtlasHomePageSpeciesCard(PopularSpeciesInfo popularSpeciesInfo) {
+        List content = new ArrayList();
```
In Java you specify the type when you instantiate a new list. So it should look something like `List<Pair<String, Optional<String>>> content = new ArrayList<>()`. Or you could use `Arrays.asList()` rather than repeatedly adding to the `content` variable, see here: https://www.geeksforgeeks.org/arrays-aslist-method-in-java-with-examples/

monicajianu commented on base/src/main/java/uk/ac/ebi/atlas/model/card/CardModelFactory.java:
```diff
@@ -32,6 +33,24 @@ public CardModel create(PopularSpeciesInfo popularSpeciesInfo) {
     }
 
+
+    public CardModel createAtlasHomePageSpeciesCard(PopularSpeciesInfo popularSpeciesInfo) {
```
All in all, assuming `ATLASSPECIES` is no different to `SPECIES`, this method is very similar to the `create()` method above. Is there any way you could refactor to minimise code duplication?

monicajianu commented on base/src/main/java/uk/ac/ebi/atlas/species/AtlasInformationDao.java:
```diff
@@ -0,0 +1,62 @@
+        return atlasInformation;
+    }
+
+    private ArrayList<String> readAtlasInformation(JsonReader reader) throws IOException {
```
The structure of the `atlas-information.json` needs to be more well-defined. The keys should describe the value stored in the field, rather than where the information is displayed on the page. What would happen if we provided this file for download, and a user wanted to find out the Ensembl Genomes version was in a programmatic way? They would have to parse the string stored in the `right` field and they would have to make a lot of assumptions.

Also, you should retrieve the number of species, assays and studies by querying the database. This information will not be updated by data production, as we have all the information we require.

Additionally, when you define a new schema for a file that will have to be used/updated by another team, you should document the schema somewhere on Confluence. That way everyone is clear about the expected file structure.

lingyun1010 replied:
I have changed the json file as following.
```
[{
  "Ensembl": 94, 
  "Ensembl Genomes": 41, 
  "WormBase paraSite": 11, 
  "EFO": 2.98
}]
```

monicajianu replied:
Much better! However, while there isn't an established standard for naming JSON keys, most conventions dictate that you shouldn't use spaces in the keys. A lot of parsers can't really handle them. Remember also that this file will be updated programmatically, as part of a pipeline. We want to make it as easy to parse as possible.

Have a look at other JSON objects we produce, either through endpoints or in files (for example, `species-properties.json`). Can you spot a naming convention for the keys?

Also, your object is wrapped in an array. Is that necessary? Do you envision other objects being stored in this file?

lingyun1010 replied:
Yes. Thank you for your great comments. I have changed the json file into 
```
{
  "ensembl": 94, 
  "genomes": 41, 
  "paraSite": 11, 
  "efo": 2.98
}
```

monicajianu commented on base/src/test/java/uk/ac/ebi/atlas/model/card/CardModelFactoryTest.java:
```diff
@@ -67,6 +70,26 @@ void createPopularSpeciesCard() {
     }
 
+    @Test
+    void createAtlasPopularSpeciesCard() {
```
This test is almost exactly the same as the one above. We should aim to minimise code duplication.

monicajianu commented on gxa/src/main/webapp/WEB-INF/jsp/tiles/views/home/browse-by.jsp:
```diff
@@ -1,73 +1,22 @@
-</div>
-
-<script defer src="${pageContext.request.contextPath}/resources/js-bundles/expressionAtlasBrowseBySpecies.bundle.js"></script>
+<link rel="stylesheet" href="https://ebi.emblstatic.net/web_guidelines/EBI-Framework/v1.3/css/ebi-global.css" type="text/css" media="all" />
```
These stylesheets have already been added in the `base.jsp` file, so they shouldn't be added here as well.

monicajianu commented on gxa/src/main/webapp/WEB-INF/jsp/tiles/views/home/experiment-list-latest.jsp:
```diff
@@ -13,58 +12,56 @@
+                <div class="small-2 columns combo text-center padding-top-medium">
                     <a href="${pageContext.request.contextPath}/experiments?experimentSet=ENCODE">
-                        <img class="margin-bottom-large experiment-thumbnail" src="${pageContext.request.contextPath}/resources/images/experiment-list-latest/encode.png" />
+                        <img class="margin-bottom-large experiment-thumbnail" style="height: 100px;" src="${pageContext.request.contextPath}/resources/images/experiment-list-latest/encode.png" />
```
The featured experiment thumbnail are really too large, they should be a bit smaller. As it is they are taking over the page. Also, when you hover over some of them you can see a thick black line (see screenshot).
![screenshot 2019-01-17 at 10 52 35](https://user-images.githubusercontent.com/5923538/51313764-476efe00-1a46-11e9-9405-d8674f540c3c.png)

It would also be nice if the thumbnails and the container were responsive. When you shrink the page, some of them appear rather squished:
![screenshot 2019-01-17 at 10 55 42](https://user-images.githubusercontent.com/5923538/51313869-8dc45d00-1a46-11e9-84d2-7f6c756f8a02.png)

Maybe 2 rows of 6 elements is not the best approach if the screen size is smaller?

lingyun1010 replied:
Nice point. I have changed it to 3 elements when the screen is small.

monicajianu commented on sc/src/main/javascript/bundles/scxa-homepage-panels/html/index.html:
```diff
@@ -24,15 +24,18 @@
 <!-- Set to http://localhost:8080/gxa/ or http://localhost:8080/gxa_sc/ -- Remember the trailing slash! -->
```
This should not be part of this PR. Please remove it and all SC-related code and make a separate PR for changes related to Single Cell.

monicajianu reviewed (changes requested):
Good start, a few changes are required though.

lingyun1010 committed:
Refactor Cardmodel create methods for SPECIES and

 ATLASSPECIES types

lingyun1010 committed:
Refactor file reader for atlas information

lingyun1010 committed:
Pass the number of assays, species studies to home

lingyun1010 committed:
Add atlas information to home page

lingyun1010 committed:
Remove duplicate  css style

lingyun1010 committed:
Redesign featured experiment css in homepage

lingyun1010 committed:
Revert "Update props for homepage panel container in SC"

This reverts commit 1235c700382febaaa28ea0a0308bb5668256a35b.

lingyun1010 committed:
Read json file has a single object with simple …

object names

lingyun1010 committed:
Refactor duplicated code

lingyun1010 committed:
Remove ATLASSPECIES cardtype and renew SPECIES

monicajianu commented on gxa/src/main/java/uk/ac/ebi/atlas/home/HomeController.java:
```diff
@@ -76,17 +103,27 @@ public String getHome(Model model) {
-
         model.addAttribute("speciesPath", ""); // Required by Spring form tag
 
+        JsonArray experimentsjsonArray = experimentInfoListService.getExperimentsJson().get("aaData").getAsJsonArray();
```
I think you could get data straight from services and DAO classes, as opposed to parsing the JSON in the backend. If you think about it, you're going from Java objects -> JSON -> Java objects. It would make your life much easier if you skipped the JSON parsing in the middle! In order to do this, you could make use of the  `listPublicExperiments()` method in the `ExperimentInfoListService`, for example.

monicajianu commented on base/src/main/java/uk/ac/ebi/atlas/species/AtlasInformationDao.java:
```diff
@@ -0,0 +1,44 @@
+            while (reader.hasNext()) {
+                String name = reader.nextName();
+                switch(name){
+                    case "ensembl": atlasInformation.add("Ensembl: " + reader.nextString()); break;
```
It is not the DAO's responsibility to format the data, its only responsibility is to retrieve it. The UI is responsible for saying what this data means. The side-effect of this implementation is that you end up doing strange things in the `.jsp` to add commas. What would you do if you wanted to display the EFO version, for example, on a different page altogether? You'd have to parse each string in the list and find out which one is the EFO version, then extract the number.

Can you think of a better data structure that would allow you to retrieve the value for any given piece of metadata? Also, can you think of a way to identify the data without having to pass strings around? Hint: read about [enums in Java](https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html).

lingyun1010 replied:
Thank you for your comments. I have changed the data structure into Hashmap to retrieve any piece of metadata. However, I did not see the point of `enums` usage here, since the data we read from the file is the type of `String`, why we do not just pass strings around directly?

monicajianu replied:
What happens if we change the structure of the file? Let's say we want to rename "paraSite" to "Fred". Let's also imagine that you used the string "paraSite" to extract the information from the map in 3 different classes. You'll end up having to change 3 different files to change the property name. However, if you use enums, you only need to change 1 file, i.e. the definition of the enum. An enum would also allow you to make use of more advanced features of your IDE to find out where they're used. If you use strings, your only option is using the Find function to search across all the files in the project. 

You're probably thinking, "well I know exactly which files need changing, I won't need to use Find anyway". But what if 6 months have gone by? What if it's not you changing the files, but another developer?

lingyun1010 replied:
Wow, I see. I like your "probably thinking" paragraph. :) I will have a corresponding change. Thank you for your detailed description.

monicajianu commented on base/src/main/java/uk/ac/ebi/atlas/species/AtlasInformationDao.java:
```diff
@@ -0,0 +1,44 @@
+                        atlasInformationFilePath.getFileName().toString());
+    }
+
+    public ArrayList<String> fetchAll() throws IOException {
```
I know you modelled this on `SpeciesPropertiesDao`, but I think the JSON parsing logic could be simplified a LOT if you used [Gson](https://google.github.io/gson/apidocs/com/google/gson/Gson.html#fromJson-java.io.Reader-java.lang.Class-). So your code could be as simple as:
```
 JsonReader reader = new JsonReader(new FileReader(atlasInformationFilePath.toString()));
 SomeClass data = GsonProvider.GSON.fromJson(reader, SomeCLass.class);
```

@alfonsomunozpomer Is there are a reason why the `SpeciesPropertiesDao` is so convoluted?

monicajianu commented on base/src/main/java/uk/ac/ebi/atlas/model/card/CardModelFactory.java:
```diff
@@ -21,15 +22,29 @@ public CardModelFactory(UrlHelpers urlHelpers) {
+                Optional.of(urlHelpers.getExperimentsFilteredBySpeciesAndExperimentType(popularSpeciesInfo.species(), "baseline"))));
+        }
+        if(popularSpeciesInfo.differentialExperiments() > 0){
+            content.add(Pair.of("Differential: " + popularSpeciesInfo.differentialExperiments() + " ",
```
You don't need the blank space you append at the end of this string.

monicajianu commented on base/src/main/java/uk/ac/ebi/atlas/model/card/CardModelFactory.java:
```diff
@@ -21,15 +22,29 @@ public CardModelFactory(UrlHelpers urlHelpers) {
+                Optional.of(urlHelpers.getExperimentsFilteredBySpeciesAndExperimentType(popularSpeciesInfo.species(), "differential"))));
+        }
+
+        return createCardModel(popularSpeciesInfo, CardIconType.SPECIES, content);
```
You only call this private `createCardModel` method once. Was the intention to call it from `createLandingPageImageCard` as well?

monicajianu reviewed (changes requested):
I added a few more comments. Don't forget to run Checkstyle on the code to spot formatting issues. Also, keep an eye on Bamboo! There are a few failing tests for this branch.

lingyun1010 committed:
Rename the json file

lingyun1010 committed:
Reformat code and delete whitespace

lingyun1010 committed:
Simplify AtlasInformationDao

lingyun1010 committed:
Change the data structure and remove comma editing

lingyun1010 committed:
Change test for new SPECIES card type

lingyun1010 committed:
Add a new developer ;)

monicajianu commented on gxa/src/main/java/uk/ac/ebi/atlas/home/HomeController.java:
```diff
@@ -76,17 +104,34 @@ public String getHome(Model model) {
 
+        List<ExperimentInfo> experimentlist = experimentInfoListService.listPublicExperiments();
+
+        Double numberOfAssays = 0.0;
```
Why is `numberOfAssays` a `Double`, which you then round later on?

lingyun1010 replied:
The data I retrieved from experimentInfoListSevice is `Double`. I want to get an integer then.

monicajianu replied:
I can't see any `Double`s in the `ExperimentInfo` class. 

![screenshot 2019-01-31 at 13 01 51](https://user-images.githubusercontent.com/5923538/52055843-731ad980-2558-11e9-9602-06301b1e42c7.png)

monicajianu commented on gxa/src/main/java/uk/ac/ebi/atlas/home/HomeController.java:
```diff
@@ -76,17 +104,34 @@ public String getHome(Model model) {
+                .replace("genomes", "Ensembl Genomes")
+                .replace("paraSite", "WormBase paraSite")
+                .replace("efo", "EFO")
+                .replace("=",":")
```
If you find yourself doing any kind of manipulation on a string, ask yourself if there are better approaches for what you are trying to do. It is not necessary to build the sentence "Ensemble x, Esembl Genomes y, etc etc" in the controller. Why can't you use the same clean approach as you did to display the number of studies and assays?

lingyun1010 committed:
Add an enum data type for atlas information

lingyun1010 committed:
Add dataType from enum as model attribute

lingyun1010 committed:
Correct the numberOfAssays variable type to int

lingyun1010 committed:
Debug css style about text overlapping when resize

the window

lingyun1010 committed:
Correct typo

lingyun1010 committed:
Rewrite data in json file, clearer for curators

{
  "ensembl": "94",
  "ensembl_genomes": "41",
  "wormbase_parasite": "11",
  "efo": "2.98"
}

lingyun1010 committed:
Sync the css style as the original design in slack

monicajianu reviewed (approved):
Looks good! Great work! 😃
